data/src/main/java/com/microsoft/azure/kusto/data/auth/TokenProviderBase.java

```diff
@@ -24,7 +25,10 @@
         this.clusterUrl = UriUtils.setPathForUri(clusterUrl, "");
     }
 
-    protected void initializeCloudInfo() throws DataServiceException {
+    protected void setRequiredMembersBasedOnCloudInfo() throws DataClientException, DataServiceException {
+    }
+
+    protected void initializeCloudInfo() throws DataClientException, DataServiceException {

```

Comment:
Why is this not part of TokenProviderFactory? The class is not usable without CloudInfo so for me this is a perfect match for a factory.
I would change the C'tor to accept CloudInfo instead of clusterUrl and move setRequiredMembersBasedOnCloudInfo logic to the C'tor.

Reply:
Some Token Providers don't need cloud info - AccessTokenTokenProvider, CallbackTokenProvider so we don't need to do the web request every time.

And we do use clusterUrl in the derived classes, to give more info in the exception.

Reply:
"Some Token Providers don't need cloud info" and yet you have two functions (setRequiredMembersBasedOnCloudInfo,initializeCloudInfo) with could info in their name.
Also you have a function in the base class  (determineScope) which depends on cloud info.
I don't understand the design of this class... We can talk about in Teams if you would like.